### ISSUE TITLE
Use of amp.autocast does not improve performance 

### ISSUE THREAD

**Issue by aurotripathy** [karpathy/minGPT#51]
I'm experimenting with  `amp.autocast` (automatic mixed precision) with  torch version, '1.7.0a0+7036e91'
I find the performance has _not improved_ (slight degraded).
Changes in bold and they follow the [amp recipe](https://pytorch.org/tutorials/recipes/recipes/amp_recipe.html)

<pre>
    def train(self):
        <b>use_amp = self.config.precision == 'AMP'
        scaler = torch.cuda.amp.GradScaler(enabled=use_amp)  # Setup once at the beginning of training</b>

        model, config = self.model, self.config
        raw_model = model.module if hasattr(self.model, "module") else model
        optimizer = raw_model.configure_optimizers(config)

        def run_epoch(split):
            is_train = split == 'train'
            model.train(is_train)
            data = self.train_dataset if is_train else self.test_dataset
            loader = DataLoader(data, shuffle=True, pin_memory=True,
                                batch_size=config.batch_size,
                                num_workers=config.num_workers)

            losses = []
            pbar = tqdm(enumerate(loader), total=len(loader)) if is_train else enumerate(loader)
            for it, (x, y) in pbar:

                # place data on the correct device
                x = x.to(self.device)
                y = y.to(self.device)

                # forward the model
                with torch.set_grad_enabled(is_train):
                    <b>with torch.cuda.amp.autocast(enabled=use_amp):  # cast ops in mixed precision
                            logits, loss = model(x, y)</b>
                    loss = loss.mean() # collapse all losses if they are scattered on multiple gpus
                    losses.append(loss.item())

                if is_train:
                    model.zero_grad()
                    <b># Scale loss, call backward() to create scaled gradients
                    scaler.scale(loss).backward()
                    # unscale gradient, call optimizer.step()
                    scaler.step(optimizer)
                    # Update the scale for next iteration
                    scaler.update()</b>
                    # TBD deleted 'torch.nn.utils.clip_grad_norm_(model.parameters(), config.grad_norm_clip)'

</pre>

**Comment by MiaozhiYu** [comment]
i have the same problem. basically the same code as you
